File: app/src/main/java/dev/favourdevlabs/cleanthes/autofill/AutofillAuthActivity.java
package dev.favourdevlabs.cleanthes.autofill;

import android.content.Intent;
import android.os.Bundle;
import android.service.autofill.Dataset;
import android.service.autofill.FillResponse;
import android.view.WindowManager;
import android.view.autofill.AutofillId;
import android.view.autofill.AutofillManager;
import android.view.autofill.AutofillValue;
import android.widget.RemoteViews;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;

import dev.favourdevlabs.cleanthes.R;
import dev.favourdevlabs.cleanthes.data.entities.VaultEntry;
import dev.favourdevlabs.cleanthes.data.repository.VaultRepository;
import dev.favourdevlabs.cleanthes.ui.auth.SessionManager;

import java.util.ArrayList;
import java.util.List;
import javax.crypto.SecretKey;

public class AutofillAuthActivity extends AppCompatActivity {

    public static final String EXTRA_PACKAGE_NAME = "pkg";
    public static final String EXTRA_WEB_DOMAIN   = "domain";
    public static final String EXTRA_USERNAME_ID  = "uid";
    public static final String EXTRA_PASSWORD_ID  = "pid";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_autofill_auth);

        getWindow().setFlags(
                WindowManager.LayoutParams.FLAG_SECURE,
                WindowManager.LayoutParams.FLAG_SECURE
        );

        // Vault is locked — user must open Cleanthes and unlock first
        if (!SessionManager.isUnlocked()) {
            setResult(RESULT_CANCELED);
            finish();
            return;
        }

        prompt();
    }

    private void prompt() {
        new BiometricPrompt(this, ContextCompat.getMainExecutor(this),
                new BiometricPrompt.AuthenticationCallback() {

                    @Override
                    public void onAuthenticationSucceeded(
                            @NonNull BiometricPrompt.AuthenticationResult result) {
                        deliver();
                    }

                    @Override
                    public void onAuthenticationFailed() {}

                    @Override
                    public void onAuthenticationError(
                            int code, @NonNull CharSequence msg) {
                        setResult(RESULT_CANCELED);
                        finish();
                    }
                }
        ).authenticate(new BiometricPrompt.PromptInfo.Builder()
                .setTitle("Cleanthes")
                .setSubtitle("Authenticate to fill")
                .setNegativeButtonText("Cancel")
                .build());
    }

    private void deliver() {
        SecretKey secretKey = SessionManager.getSessionKey();
        if (secretKey == null) {
            setResult(RESULT_CANCELED);
            finish();
            return;
        }

        Intent in = getIntent();
        AutofillId usernameId = in.getParcelableExtra(EXTRA_USERNAME_ID);
        AutofillId passwordId = in.getParcelableExtra(EXTRA_PASSWORD_ID);
        String packageName    = in.getStringExtra(EXTRA_PACKAGE_NAME);
        String webDomain      = in.getStringExtra(EXTRA_WEB_DOMAIN);
        String lookupKey      = webDomain != null ? webDomain : packageName;

        try {
            List<VaultEntry> all = VaultRepository.getInstance(this).getAllEntries(secretKey);
            List<VaultEntry> matches = filter(all, lookupKey);

            if (matches.isEmpty()) {
                setResult(RESULT_CANCELED);
                finish();
                return;
            }

            FillResponse.Builder response = new FillResponse.Builder();
            for (VaultEntry entry : matches) {
                RemoteViews view = new RemoteViews(getPackageName(), R.layout.autofill_item);
                view.setTextViewText(R.id.autofill_label, entry.getUsername());

                // After getAllEntries(), encryptedPassword field holds the
                // decrypted plain password — VaultRepository.decryptEntry()
                // overwrites the field in place.
                response.addDataset(new Dataset.Builder(view)
                        .setValue(usernameId,
                                AutofillValue.forText(entry.getUsername()), view)
                        .setValue(passwordId,
                                AutofillValue.forText(entry.getEncryptedPassword()), view)
                        .build());
            }

            SessionManager.refreshSession();

            Intent out = new Intent();
            out.putExtra(AutofillManager.EXTRA_AUTHENTICATION_RESULT, response.build());
            setResult(RESULT_OK, out);

        } catch (Exception e) {
            setResult(RESULT_CANCELED);
        }

        finish();
    }

    private List<VaultEntry> filter(List<VaultEntry> entries, String key) {
        List<VaultEntry> out = new ArrayList<>();
        if (key == null || key.isEmpty()) return out;
        String lower = key.toLowerCase();
        for (VaultEntry e : entries) {
            String website = e.getWebsite() != null ? e.getWebsite().toLowerCase() : "";
            String title   = e.getTitle()   != null ? e.getTitle().toLowerCase()   : "";
            if (website.contains(lower) || lower.contains(website)
                    || title.contains(lower)) {
                out.add(e);
            }
        }
        return out;
    }
}

File: feature/autofill/src/main/java/dev/favourdevlabs/cleanthes/feature/autofill/AutofillAuthActivity.kt
package dev.favourdevlabs.cleanthes.feature.autofill

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.service.autofill.FillResponse
import android.util.Log
import android.view.autofill.AutofillId
import android.view.autofill.AutofillManager
import androidx.activity.compose.setContent
import androidx.biometric.BiometricPrompt
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.ui.Modifier
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import dev.favourdevlabs.cleanthes.domain.model.CitadelItem
import dev.favourdevlabs.cleanthes.data.api.CitadelRepository
import dev.favourdevlabs.cleanthes.security.session.SessionManager
import dev.favourdevlabs.cleanthes.ui.base.SecureActivity
import dev.favourdevlabs.cleanthes.ui.theme.CleanthesTheme
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.URI
import javax.inject.Inject

@AndroidEntryPoint
class AutofillAuthActivity : SecureActivity() {
    @Inject lateinit var sessionManager: SessionManager

    @Inject lateinit var repository: CitadelRepository

    companion object {
        private const val TAG = "AutofillAuthActivity"
        const val EXTRA_PACKAGE_NAME = "pkg"
        const val EXTRA_WEB_DOMAIN = "domain"
        const val EXTRA_USERNAME_ID = "uid"
        const val EXTRA_PASSWORD_ID = "pid"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            CleanthesTheme {
                Box(
                    modifier =
                        Modifier
                            .fillMaxSize()
                            .background(MaterialTheme.colorScheme.background),
                )
            }
        }
        if (sessionManager.lockState.value) {
            setResult(RESULT_CANCELED)
            finish()
            return
        }
        prompt()
    }

    private fun prompt() {
        BiometricPrompt(
            this,
            ContextCompat.getMainExecutor(this),
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) = deliver()

                override fun onAuthenticationFailed() = Unit

                override fun onAuthenticationError(
                    code: Int,
                    msg: CharSequence,
                ) {
                    Log.w(TAG, "biometric auth error: $code $msg")
                    setResult(RESULT_CANCELED)
                    finish()
                }
            },
        ).authenticate(
            BiometricPrompt.PromptInfo
                .Builder()
                .setTitle("Cleanthes")
                .setSubtitle("Authenticate to fill")
                .setNegativeButtonText("Cancel")
                .build(),
        )
    }

    private fun deliver() {
        val secretKey =
            sessionManager.getSessionKey()
                ?: run {
                    Log.w(TAG, "deliver: no session key, session locked or expired")
                    setResult(RESULT_CANCELED)
                    finish()
                    return
                }
        val usernameId = getParcelableExtraCompat<AutofillId>(EXTRA_USERNAME_ID)
        val passwordId = getParcelableExtraCompat<AutofillId>(EXTRA_PASSWORD_ID)
        val packageName = intent.getStringExtra(EXTRA_PACKAGE_NAME)
        val webDomain = intent.getStringExtra(EXTRA_WEB_DOMAIN)
        val lookupKey = webDomain ?: packageName

        if (usernameId == null || passwordId == null) {
            Log.w(TAG, "deliver: missing usernameId/passwordId in auth intent")
            setResult(RESULT_CANCELED)
            finish()
            return
        }

        lifecycleScope.launch {
            try {
                val matches =
                    if (lookupKey.isNullOrEmpty()) {
                        emptyList()
                    } else {
                        withContext(Dispatchers.IO) {
                            filter(repository.getEntriesByDomainCandidate(lookupKey, secretKey), lookupKey)
                        }
                    }
                if (matches.isEmpty()) {
                    Log.d(TAG, "deliver: no matching entries for $lookupKey")
                    setResult(RESULT_CANCELED)
                    finish()
                    return@launch
                }
                val response = FillResponse.Builder()
                for (entry in matches) {
                    response.addDataset(
                        DatasetBuilder.build(
                            this@AutofillAuthActivity,
                            usernameId,
                            passwordId,
                            entry,
                        ),
                    )
                }
                sessionManager.refreshSession()
                setResult(
                    RESULT_OK,
                    Intent().putExtra(
                        AutofillManager.EXTRA_AUTHENTICATION_RESULT,
                        response.build(),
                    ),
                )
            } catch (e: Exception) {
                Log.w(TAG, "deliver failed", e)
                setResult(RESULT_CANCELED)
            }
            finish()
        }
    }

    private inline fun <reified T> getParcelableExtraCompat(key: String): T? =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            intent.getParcelableExtra(key, T::class.java)
        } else {
            @Suppress("DEPRECATION")
            intent.getParcelableExtra(key)
        }

    private fun filter(
        entries: List<CitadelItem>,
        key: String?,
    ): List<CitadelItem> {
        if (key.isNullOrEmpty()) return emptyList()
        val targetHost = registrableDomain(key) ?: key.lowercase()
        return entries.filter { e ->
            val entryHost = e.website?.let { registrableDomain(it) }
            val hostMatch = entryHost != null && entryHost == targetHost
            val titleMatch = e.title.lowercase() == targetHost
            hostMatch || titleMatch
        }
    }

    private fun registrableDomain(input: String): String? {
        val candidate = if (input.contains("://")) input else "https://$input"
        val host =
            try {
                URI(candidate).host
            } catch (_: Exception) {
                null
            } ?: return null
        val lower = host.lowercase()
        val parts = lower.split(".")
        return if (parts.size >= 2) parts.takeLast(2).joinToString(".") else lower
    }
}

